File: examples/C/get_vara.c
/*********************************************************************
 *
 *  Copyright (C) 2014, Northwestern University and Argonne National Laboratory
 *  See COPYRIGHT notice in top-level directory.
 *
 *********************************************************************/
/* $Id$ */

/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * This example is the read counterpart of example put_vara.c. It shows how to
 * use ncmpi_get_vara_int_all() to read a 2D 4-byte integer array in parallel.
 * It also reads a global attribute and two attribute of variable named "var".
 * The data partitioning pattern is a column-wise partitioning across all
 * processes. Each process reads a subarray of size local_ny * local_nx.
 *
 *    To compile:
 *        mpicc -O2 get_vara.c -o get_vara -lpnetcdf
 *
 * Input file is the output file produced by put_vara.c. Here is the CDL dumped
 * from running ncmpidump.
 *
 *    % ncmpidump /pvfs2/wkliao/testfile.nc
 *    netcdf testfile {
 *    // file format: CDF-1
 *    dimensions:
 *            y = 10 ;
 *            x = 16 ;
 *    variables:
 *            int var(y, x) ;
 *                var:str_att_name = "example attribute of type text." ;
 *                var:float_att_name = 0.f, 1.f, 2.f, 3.f, 4.f, 5.f, 6.f, 7.f ;
 *                var:short_att_name = 1000s ;
 *    // global attributes:
 *                :history = "Mon Aug 13 21:27:48 2018" ;
 *       "" ;
 *    data:
 *
 *     var =
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
 *         0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3 ;
 *    }
 *
 * Example command for MPI run:
 *
 *    % mpiexec -n 4 ./get_vara /pvfs2/wkliao/testfile.nc
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#include <stdio.h>
#include <stdlib.h>
#include <string.h> /* strcpy(), strncpy() */
#include <unistd.h> /* getopt() */
#include <mpi.h>
#include <pnetcdf.h>

static int verbose;

#define ERR {if(err!=NC_NOERR){printf("Error at %s:%d : %s\n", __FILE__,__LINE__, ncmpi_strerror(err));return 1;}}

static void
usage(char *argv0)
{
    char *help =
    "Usage: %s [-h] | [-q] [file_name]\n"
    "       [-h] Print help\n"
    "       [-q] Quiet mode (reports when fail)\n"
    "       [filename] input netCDF file name\n";
    fprintf(stderr, help, argv0);
}

/*----< pnetcdf_check_mem_usage() >------------------------------------------*/
/* check PnetCDF library internal memory usage */
static int
pnetcdf_check_mem_usage(MPI_Comm comm)
{
    int err, nerrs=0, rank;
    MPI_Offset malloc_size, sum_size;

    MPI_Comm_rank(comm, &rank);

    /* print info about PnetCDF internal malloc usage */
    err = ncmpi_inq_malloc_max_size(&malloc_size);
    if (err == NC_NOERR) {
        MPI_Reduce(&malloc_size, &sum_size, 1, MPI_OFFSET, MPI_SUM, 0, MPI_COMM_WORLD);
        if (rank == 0 && verbose)
            printf("maximum heap memory allocated by PnetCDF internally is %lld bytes\n",
                   sum_size);

        /* check if there is any PnetCDF internal malloc residue */
        err = ncmpi_inq_malloc_size(&malloc_size);
        MPI_Reduce(&malloc_size, &sum_size, 1, MPI_OFFSET, MPI_SUM, 0, MPI_COMM_WORLD);
        if (rank == 0 && sum_size > 0)
            printf("heap memory allocated by PnetCDF internally has %lld bytes yet to be freed\n",
                   sum_size);
    }
    else if (err != NC_ENOTENABLED) {
        printf("Error at %s:%d: %s\n", __FILE__,__LINE__,ncmpi_strerror(err));
        nerrs++;
    }
    return nerrs;
}

/*----< pnetcdf_io() >-------------------------------------------------------*/
static int
pnetcdf_io(MPI_Comm comm, char *filename)
{
    char str_att[NC_MAX_NAME];
    int rank, nprocs, err, nerrs=0, ncid, varid, dimid[2], *buf;
    float *float_att;
    MPI_Offset len, global_ny, global_nx, local_ny, local_nx;
    MPI_Offset start[2], count[2];

    MPI_Comm_rank(comm, &rank);
    MPI_Comm_size(comm, &nprocs);

    /* open an existing file for reading -------------------------------------*/
    err = ncmpi_open(comm, filename, NC_NOWRITE, MPI_INFO_NULL, &ncid); ERR

    /* get global attribute named "history" */
    err = ncmpi_get_att_text(ncid, NC_GLOBAL, "history", str_att); ERR
    err = ncmpi_inq_attlen(ncid, NC_GLOBAL, "history", &len); ERR
    str_att[len] = '\0'; /* add a NULL char at the end */
    if (rank == 0 && verbose)
        printf("global attribute \"history\" of text: %s\n",str_att);

    /* get dimension IDs for dimensions Y and X */
    err = ncmpi_inq_dimid(ncid, "Y", &dimid[0]); ERR
    err = ncmpi_inq_dimid(ncid, "X", &dimid[1]); ERR

    /* get dimension lengths for dimensions Y and X */
    err = ncmpi_inq_dimlen(ncid, dimid[0], &global_ny); ERR
    err = ncmpi_inq_dimlen(ncid, dimid[1], &global_nx); ERR

    /* get the variable ID of a 2D variable of integer type */
    err = ncmpi_inq_varid(ncid, "var", &varid); ERR

    /* get variable's attribute named "str_att_name" */
    err = ncmpi_get_att_text(ncid, varid, "str_att_name", str_att); ERR
    err = ncmpi_inq_attlen(ncid, varid, "str_att_name", &len); ERR
    str_att[len] = '\0'; /* add a NULL char at the end */
    if (rank == 0 && verbose)
        printf("variable attribute \"str_att_name\" of type text = \"%s\"\n",
               str_att);

    /* get the length of variable's attribute named "float_att_name" */
    err = ncmpi_inq_attlen(ncid, varid, "float_att_name", &len); ERR

    /* get attribute contents */
    float_att = (float*) malloc(len * sizeof(float));
    err = ncmpi_get_att_float(ncid, varid, "float_att_name", float_att); ERR
    free(float_att);

    /* the local array size */
    local_ny = global_ny;
    local_nx = global_nx / nprocs;
    buf = (int*) malloc(local_nx * local_ny * sizeof(int));

    /* prepare reading subarray */
    start[0] = 0;
    start[1] = local_nx * rank;
    count[0] = local_ny;
    count[1] = local_nx;

    /* read a subarray in collective mode */
    err = ncmpi_get_vara_int_all(ncid, varid, start, count, buf); ERR
    free(buf);

    err = ncmpi_close(ncid); ERR

    return nerrs;
}

int main(int argc, char** argv)
{
    extern int optind;
    char filename[256];
    int i, rank, nerrs=0;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    verbose = 1;

    /* get command-line arguments */
    while ((i = getopt(argc, argv, "hq")) != EOF)
        switch(i) {
            case 'q': verbose = 0;
                      break;
            case 'h':
            default:  if (rank==0) usage(argv[0]);
                      MPI_Finalize();
                      return 1;
        }
    if (argv[optind] == NULL) strcpy(filename, "testfile.nc");
    else                      snprintf(filename, 256, "%s", argv[optind]);

    nerrs += pnetcdf_io(MPI_COMM_WORLD, filename);

    nerrs += pnetcdf_check_mem_usage(MPI_COMM_WORLD);

    MPI_Finalize();
    return (nerrs > 0);
}

File: examples/F77/get_vara.f
!
!   Copyright (C) 2013, Northwestern University
!   See COPYRIGHT notice in top-level directory.
!
! $Id$

!
! This example shows how to use nfmpi_get_vara_int_all() to read a 2D
! 4-byte integer array in parallel. It is the opposite of program
! 'put_vara.f'. It reads a netCDF variable defined in the input file of
! size global_nx * global_ny where
!    global_nx == 5 and
!    global_ny == (4 * number of MPI processes).
! The data partitioning pattern is a column-wise partitioning across all
! processes. Each process writes a subarray of size nx * ny.
! Note the description above follows the Fortran array index order.
!
! Example commands for MPI run.
!
!    % mpif77 -O2 -o get_vara get_vara.f -lpnetcdf
!    % mpiexec -n 4 ./get_vara /pvfs2/wkliao/testfile.nc
!
      subroutine check(err, message)
          implicit none
          include 'mpif.h'
          include 'pnetcdf.inc'
          integer err
          character message*(*)

          ! It is a good idea to check returned value for possible error
          if (err .NE. NF_NOERR) then
              write(6,*) message//' '//nfmpi_strerror(err)
              call MPI_Abort(MPI_COMM_WORLD, -1, err)
          end if
      end ! subroutine check

      program main
          implicit none
          include 'mpif.h'
          include 'pnetcdf.inc'

          character(LEN=256) filename, cmd
          integer i, j, err, ierr, nprocs, rank, get_args, dummy
          integer cmode, ncid, varid, dimid(2)
          integer*8 nx, ny, global_nx, global_ny
          integer*8 starts(2), counts(2)
          PARAMETER(nx=5, ny=4)
          integer buf(nx,ny)
          integer*8 malloc_size, sum_size
          logical verbose

          call MPI_Init(err)
          call MPI_Comm_rank(MPI_COMM_WORLD, rank, err)
          call MPI_Comm_size(MPI_COMM_WORLD, nprocs, err)

          ! take filename from command-line argument if there is any
          if (rank .EQ. 0) then
              verbose = .TRUE.
              filename = "testfile.nc"
              ierr = get_args(2, cmd, filename, verbose, dummy)
          endif
          call MPI_Bcast(ierr, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, err)
          if (ierr .EQ. 0) goto 999

          call MPI_Bcast(verbose, 1, MPI_LOGICAL, 0, MPI_COMM_WORLD,
     +                   err)
          call MPI_Bcast(filename, 256, MPI_CHARACTER, 0,
     +                   MPI_COMM_WORLD, err)

          ! open file
          err = nfmpi_open(MPI_COMM_WORLD, filename, NF_NOWRITE,
     +                     MPI_INFO_NULL, ncid)
          call check(err, 'In nfmpi_open: ')

          ! inquire dimensions x and y
          err = nfmpi_inq_dimid(ncid, "x", dimid(1))
          call check(err, 'In nfmpi_inq_dim x: ')
          err = nfmpi_inq_dimid(ncid, "y", dimid(2))
          call check(err, 'In nfmpi_inq_dim y: ')

          ! get dimension lengths for dimensions Y and X */
          err = nfmpi_inq_dimlen(ncid, dimid(1), global_nx)
          call check(err, 'In nfmpi_inq_dimlen x: ')
          err = nfmpi_inq_dimlen(ncid, dimid(2), global_ny)
          call check(err, 'In nfmpi_inq_dimlen y: ')

          ! assert local array sizes
          if (global_nx .NE. nx) then
              print *, "Expect x dimension of size ",nx,
     +                  " but got ",global_nx
              call MPI_Abort(MPI_COMM_WORLD, -1, err)
          endif
          if (global_ny .NE. ny * nprocs) then
              print *, "Expect y dimension of size ",ny*nprocs,
     +                 " but got ",global_ny
              call MPI_Abort(MPI_COMM_WORLD, -1, err)
          endif

          ! get the variable ID
          err = nfmpi_inq_varid(ncid, "var", varid)
          call check(err, 'In nfmpi_inq_varid: ')

          do i=1, ny
          do j=1, nx
             buf(j,i) = -1
          enddo
          enddo

          ! Note that in Fortran, array indices start with 1
          starts(1) = 1
          starts(2) = ny * rank + 1
          counts(1) = nx
          counts(2) = ny

          err = nfmpi_get_vara_int_all(ncid, varid, starts, counts, buf)
          call check(err, 'In nfmpi_get_vara_int_all: ')

          ! close the file
          err = nfmpi_close(ncid)
          call check(err, 'In nfmpi_close: ')

          ! check contents of read buffer
 997      format(A,I4,A,I4,A,I4,A,I4)
          do i=1, ny
          do j=1, nx
             if (buf(j,i) .NE. rank) then
                 print 997,"expect read buf(",j,",",i,") to be ",rank,
     +                     " but got ",buf(j,i)
                 stop 2
             end if
          enddo
          enddo

          ! check if there is any PnetCDF internal malloc residue
 998      format(A,I13,A)
          err = nfmpi_inq_malloc_size(malloc_size)
          if (err .EQ. NF_NOERR) then
              call MPI_Reduce(malloc_size, sum_size, 1, MPI_INTEGER8,
     +                        MPI_SUM, 0, MPI_COMM_WORLD, err)
              if (rank .EQ. 0 .AND. sum_size .GT. 0)
     +            print 998,
     +            'heap memory allocated by PnetCDF internally has ',
     +            sum_size/1048576, ' MiB yet to be freed'
          endif

 999      call MPI_Finalize(err)
          ! call EXIT(0) ! EXIT() is a GNU extension
      end ! program main

